### PR TITLE
Add Tall Box simulation to unit tests and regression tests

### PR DESCRIPTION
**Pull request by chongchonghe**
### Description
This PR adds a new test problem, the Tall Box simulation. It has self-gravity, self-consistent (stochastic) star formation, supernova feedback, shielded Grackle cooling, time-variable photoelectric heating, and static mesh refinement. This test problem should also test the reproducibility of these modules: HD + gravity + particles (SF, SN) + Grackle cooling + AMR

### Related issues
None. 

### Checklist
_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_
- [ ] I have added a description (see above).
- [ ] I have added a link to any related issues (if applicable; see above).
- [ ] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [ ] I have added tests for any new physics that this PR adds to the code.
- [ ] *(For quokka-astro org members)* I have manually triggered the GPU tests with the magic comment `/azp run`.

### CODE REVIEW

File: inputs/TallBoxSf_regression.in [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+# VERBOSITY
+# *****************************************************************
+amr.v              = 1      # verbosity in Amr
+
+# *****************************************************************
+# Resolution and refinement
+# *****************************************************************
+amr.n_cell          = 64 64 256
+amr.max_level       = 2     # number of levels = max_level + 1
+amr.max_grid_size   = 64   # at least 128 for GPUs
```
[the remark this answers is not in the document]
Comment (chongchonghe) [reply inputs/TallBoxSf_regression.in]: done

**Comment by chongchonghe** [comment]
@BenWibking @markkrumholz This is ready for review.

File: extern/problems/TallBoxSf/zdrv.hdf5.zip
Comment (BenWibking) [inline extern/problems/TallBoxSf/zdrv.hdf5.zip:1]: This is a very large file. Can you try compressing with `gzip -9` instead? Or can this be generated at runtime?

File: src/simulation.hpp
Comment (BenWibking) [inline src/simulation.hpp:1]: Are these changes intended for this PR? Can you briefly explain what problem this solves?

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+
+static constexpr bool enable_self_gravity = true;
+
+constexpr double pc = C::parsec;
+constexpr double mu = 1.0 * C::m_p;
+constexpr double gamma_ = 5. / 3.;
+constexpr double arad = C::a_rad;
+constexpr double TCMB = 2.7; // K, CMB temperature
+constexpr double initial_Erad = 1e-40 * arad * TCMB * TCMB * TCMB * TCMB;
+constexpr double chat_over_c = 2000.0 * 1e5 / C::c_light; // chat = 2000 km/s
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: Does this value work in practice? It seems a bit low to me. In practice, it seems like most published simulations use 0.1*c.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+	static constexpr OpacityModel opacity_model = is_rad_on ? OpacityModel::piecewise_constant_opacity : OpacityModel::single_group;
+};
+
+template <> struct ISM_Traits<TheProblem> {
+	static constexpr bool enable_dust_gas_thermal_coupling_model = enable_dust_;
+	static constexpr bool enable_photoelectric_heating = enable_PE_;
+	static constexpr double gas_dust_coupling_threshold = 1.0e-5;
+};
+
+// Custom particle acceleration boundary condition for tall box test
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: This is something that we should really handle in AMRSimulation.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+static constexpr int nGroups_ = 1;
+static constexpr amrex::GpuArray<double, nGroups_ + 1> radBoundaries_{0.0, 100.0};
+static constexpr amrex::GpuArray<double, nGroups_ + 1> dust_opacity_{0.0, 0.0};
+// static constexpr int nGroups_ = 4;
+// static constexpr amrex::GpuArray<double, nGroups_ + 1> radBoundaries_ = { 1.e-04, 1.00778140e-01, 1.00778140e+00, 5.53817071e+00, 1.e+2 };
+// static constexpr amrex::GpuArray<double, nGroups_ + 1> dust_opacity_{6e2, 1e3, 2e4, 1e5, 2e5}; // dust opacity, cm2/g. last element not used
+static constexpr bool enable_dust_ = false;
+static constexpr bool enable_PE_ = false;
+
+static constexpr bool enable_self_gravity = true;
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: Are all of these constants outside of the Traits classes necessary?

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+	// Values in cgs units from Bate & Keto (2015), Eq. 26.
+	const double epsilon = 0.05;	   // default efficiency factor for cold molecular clouds
+	const double ref_J_ISR = 5.29e-14; // reference value for the ISR in erg cm^3
+	const double coeff = 1.33e-24;
+	return coeff * epsilon * num_density / ref_J_ISR; // s^-1
+}
+
+template <> void QuokkaSimulation<TheProblem>::createInitialStochasticStellarPopParticles()
+{
+	if (userData_.stars_file == "none") {
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: The filename could be `none.` It would be better to check for an empty string `""`.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+{
+	constexpr double gas_to_dust_ratio = 1.0e-3;
+	amrex::GpuArray<amrex::GpuArray<double, nGroups_ + 1>, 2> exponents_and_values{};
+	for (int i = 0; i < nGroups_ + 1; ++i) {
+		exponents_and_values[0][i] = 0.0; // power-law slopes
+	}
+	for (int i = 0; i < nGroups_ + 1; ++i) {
+		exponents_and_values[1][i] = dust_opacity_[i] * gas_to_dust_ratio;
+	}
+	return exponents_and_values;
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: This is a bad API design. Two different arrays shouldn't be returned like this. This needs to be fixed in a future PR so it returns either a tuple or a struct with named array members.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 7]
```diff
@@ -0,0 +1,682 @@
...
+
+	// Read particles from ASCII file. Note that this only read real components and not integer components, therefore we need to use
+	// InitSetPhyParticles to set the integer components
+	const int nreal_extra = 7; // mass vx vy vz birth_time death_time lum
+	StochasticStellarPopParticles->SetVerbose(1);
+	StochasticStellarPopParticles->InitFromAsciiFile(userData_.stars_file, nreal_extra, nullptr);
+
+	// Using a for loop from lev = 0 to StochasticStellarPopParticles->maxLevel() won't work because not all levels necessarily have particles, and when
+	// some levels do not have particles, StochasticStellarPopParticles->GetParticles(lev) will result in a Segfault. Therefore, we loop over the actual
+	// particle container.
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: This is something we need to open an AMReX issue about and reference it here.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+	amrex::ParallelFor(tags, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+		// NOTE: must check all nodes of the cell!
+		// Otherwise, cells that are too big can completely prevent refinement.
+		// amrex::Real const x0 = prob_lo[0] + (i * dx[0]);
+		// amrex::Real const y0 = prob_lo[1] + (j * dx[1]);
+		// amrex::Real const z0 = prob_lo[2] + (k * dx[2]);
+
+		// amrex::Real const x1 = prob_lo[0] + ((i + 1) * dx[0]);
+		// amrex::Real const y1 = prob_lo[1] + ((j + 1) * dx[1]);
+		// amrex::Real const z1 = prob_lo[2] + ((k + 1) * dx[2]);
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: Can you remove the commented-out code above?

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 9]
```diff
@@ -0,0 +1,682 @@
...
+template <> void QuokkaSimulation<TheProblem>::preCalculateInitialConditions()
+{
+	static bool isSamplingDone = false;
+	if (!isSamplingDone) {
+		// read perturbations from file
+		turb_data turbData;
+		amrex::ParmParse const pp("perturb");
+		std::string turbdata_filename = "zdrv.hdf5";
+		pp.query("filename", turbdata_filename);
+		initialize_turbdata(turbData, turbdata_filename);
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp:158]: Can we use the new turbulence module for this instead?

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 10]
```diff
@@ -0,0 +1,682 @@
...
+
+		// const double Tgas = P / (rho / mu * C::k_B);
+
+		const auto gamma = HydroSystem<TheProblem>::gamma_;
+
+		// add turbulent velocities
+		// Clamp indices to [turb_lo, turb_hi] range
+		// const int turb_i = amrex::Math::max(turb_lo[0], amrex::Math::min(turb_hi[0], turb_lo[0] + (i % nturb)));
+		// const int turb_j = amrex::Math::max(turb_lo[1], amrex::Math::min(turb_hi[1], turb_lo[1] + (j % nturb)));
+		// const int turb_k = amrex::Math::max(turb_lo[2], amrex::Math::min(turb_hi[2], turb_lo[2] + (k % nturb)));
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: Can you remove the commented-out code?

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 11]
```diff
@@ -0,0 +1,682 @@
...
+	const double rho_edge = consVar(i, j, kedge, HydroSystem<TheProblem>::density_index);
+	const double x1Mom_edge = consVar(i, j, kedge, HydroSystem<TheProblem>::x1Momentum_index);
+	const double x2Mom_edge = consVar(i, j, kedge, HydroSystem<TheProblem>::x2Momentum_index);
+	double x3Mom_edge = consVar(i, j, kedge, HydroSystem<TheProblem>::x3Momentum_index);
+	const double etot_edge = consVar(i, j, kedge, HydroSystem<TheProblem>::energy_index);
+	const double eint_edge = consVar(i, j, kedge, HydroSystem<TheProblem>::internalEnergy_index);
+
+	if ((x3Mom_edge * normal) < 0) { // gas is inflowing
+		x3Mom_edge *= -1.;
+	}
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: This is really wrong. You can't just change the sign here. You have to perform a geometric reflection about the boundary face and copy the corresponding ghost value. It's not just a sign-flipped foextrap.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+auto problem_main() -> int
+{
+	auto BCs_cc = quokka::BC<TheProblem>(quokka::BCType::reflecting);
+	if constexpr (BC_TYPE == 1) {
+		BCs_cc = quokka::BC<TheProblem>(quokka::BCType::int_dir, quokka::BCType::int_dir, quokka::BCType::ext_dir);
+	} else if constexpr (BC_TYPE == 2) {
+		BCs_cc = quokka::BC<TheProblem>(quokka::BCType::foextrap);
+	} else if constexpr (BC_TYPE == 3) {
+		BCs_cc = quokka::BC<TheProblem>(quokka::BCType::int_dir, quokka::BCType::int_dir, quokka::BCType::int_dir);
+	}
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: Can you use the new runtime BC infrastructure you added instead?

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 13]
```diff
@@ -0,0 +1,682 @@
...
+
+	// set random state
+	const int seed = 42;
+	amrex::InitRandom(seed, 1); // all ranks should produce the same values
+
+	// Problem initialization
+	QuokkaSimulation<TheProblem> sim(BCs_cc);
+	sim.reconstructionOrder_ = 3; // 2=PLM, 3=PPM
+	sim.cflNumber_ = 0.3;	      // *must* be less than 1/3 in 3D!
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: This can be set in the *.in file.

**Review by BenWibking — CHANGES_REQUESTED** [review]
This PR has a lot of issues.

The problem generator needs to be refactored so there aren't ad-hoc fixed for things in AMRSimulation in the problem generator (the particle acceleration BC, for instance). The particle reader bug needs to have a bug opened in the AMReX GitHub with a link in the code so that we can replace the workaround here once it's fixed in AMReX.

However, the most critical issue is that the diode BC implementation is wrong.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 9 above)
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp:158]: No. We don't need turbulence driving for this problem. Turbulence is driven by SN feedback and other stellar feedback once implemented. We only need an initial turbulence to allow stochastic star formation.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 6 above)
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: Fair point. I will do it in a future PR.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 14]
```diff
@@ -0,0 +1,685 @@
...
+	// 	kedge = khi;
+	// 	normal = 1.0;
+	// }
+
+	// This should be the correct way?
+	if (k < klo) {
+		kedge = klo;
+		normal = -1;
+	} else if (k >= khi) {
+		kedge = khi - 1;
```
[the remark this answers is not in the document]
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: @BenWibking Can you check if this comment is correct? I'm not sure if this khi is inclusive or not; More other problem generator do `k >= khi` check, so I assumed `khi` is the index of the first ghost zone.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 2 above)
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: This is a mistake. 2000 km/s is too low. I'll change it to 10000 km/s, which is 0.03 c. We definitely don't want 0.1 c; it's too high. 

It's hard to know if a given `chat` value is 'okay' or not. We need a proper test to answer this question. I recollect that the radiation shock test requires `chat = 10 cs`, and I verified `chat = 3 cs` gives the wrong answer. However, since we are relying on Grackle instead of real radiative cooling, this probably don't matter. 

With the variable reduce speed of light approximation scheme, we can probably push it further while still getting a good result.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 4 above)
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: No. I'll remove them.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 7 above)
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: Done: https://github.com/AMReX-Codes/amrex/issues/4896

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 3 above)
Comment (BenWibking) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: I thought you had a PR to handle this in AMRSimulation?

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (BenWibking) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: I will check this later today.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 11 above)
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: OK, you are right. I've made the change you suggested.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 15]
```diff
@@ -0,0 +1,537 @@
...
+	consVar(i, j, k, HydroSystem<TheProblem>::x3Momentum_index) = x3Mom;
+	consVar(i, j, k, HydroSystem<TheProblem>::energy_index) = etot;
+	consVar(i, j, k, HydroSystem<TheProblem>::internalEnergy_index) = eint;
+}
+
+auto problem_main() -> int
+{
+	// set random state
+	const int seed = 42;
+	amrex::InitRandom(seed, 1); // all ranks should produce the same values
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: I'm not sure whether we want all ranks to produce the same values. Maybe we can initialize as `seed+rank`?

**Review by BenWibking — COMMENTED** [review]
I think this looks good now, pending my double-checking of the `khi` indexing.

The only concern I have is seeding all ranks with the same random seed. I think it could produce unexpected results if we do that. For reproducibility, it might be good to set the seed to `base_seed+rank` instead.

I think we can remove the RandomBlast problem, since this tests everything it was previously testing.

**Comment by chongchonghe** [comment]
Regarding the `setFunctorParticleAccel` functor:

The issue was, when we use ext_dir on the z direction, `phi_extended.FillBoundary(geom[lev].periodicity());` does not fill in the boundary cells. `setCustomBoundaryConditions` handles the boundary cells for hydro/rad variables, but not for phi, so we have to fill in the boundary cells of phi manually. 

Somehow, not defining `setFunctorParticleAccel` only causes failure on Setonix GPUs and not on Mac. I guess it's because uninitialized arrays behaves differently on different devices.

**Comment by chongchonghe** [comment]
> I think this looks good now, pending my double-checking of the `khi` indexing.

The `khi` question is a simple one. Say, we have a box of 128 cells, and the indices goes from 0 to 127. In this case, `klo=0`, but is `khi=127` or `khi=128`? I've seen both `k > khi` and `k >= khi` when determining the upper bound cells in the problem files. I know this makes no difference in the computation because the `setCustomBoundaryConditions` function won't run through real cells at all, i.e. `if (k < klo) {} else if (k > khi) {}` is equivalent to `if (k < klo) {} else {}`, but this makes a conceptual difference, and it's important in this diode BC. 

> The only concern I have is seeding all ranks with the same random seed. I think it could produce unexpected results if we do that. For reproducibility, it might be good to set the seed to `base_seed+rank` instead.

I'll take a look. Have you done this in your galaxy run?

> I think we can remove the RandomBlast problem, since this tests everything it was previously testing.

I'll think about what to do in order to keep the total number of regression tests limited. RandomBlast problem is very helpful for debug because there are SN at the first step, whereas the tall box takes ~10 Myr to form the first star.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 15 above)
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: sounds good. done.

**Comment by chongchonghe** [comment]
> Regarding the `setFunctorParticleAccel` functor:
> 
> The issue was, when we use ext_dir on the z direction, `phi_extended.FillBoundary(geom[lev].periodicity());` does not fill in the boundary cells. `setCustomBoundaryConditions` handles the boundary cells for hydro/rad variables, but not for phi, so we have to fill in the boundary cells of phi manually.
> 
> Somehow, not defining `setFunctorParticleAccel` only causes failure on Setonix GPUs and not on Mac. I guess it's because uninitialized arrays behaves differently on different devices.

Actually, it's no longer needed somehow. I removed it from this PR and will look into it in a future PR.

**Comment by BenWibking** [comment]
Since this problem is key to figuring out the SFR issues, I am going to do a test run of this on Frontier.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: no need to check. I confirmed with a debugger that khi is inclusive, so `if (k > khi)` is the correct way.

**Comment by chongchonghe** [comment]
@BenWibking Can you review and approve this PR? I think we really need to commit to the development branch as frequent as possible to guarantee a productive workflow. If we keep a PR on hold, we have to constantly resolve merge conflicts. Plus, merging this to development branch will make running tall box sim a lot easier. 

I have removed TallBoxSf from regression and will add it once I fix the star formation issue. The tall box test problme itself is independent of the SF recipe, so I think it's okay to merge it now.

**Comment by chongchonghe** [comment]
GPU tests are not necessary since it's only merging dev branch since last pass.

File: src/problems/TallBoxSf/testTallBoxSf.cpp [hunk 16]
```diff
@@ -0,0 +1,507 @@
...
+
+	// Clamp mirror index to valid domain range for deep ghost cells
+	k_mirror = amrex::max(klo, amrex::min(khi, k_mirror));
+
+	const double rho = consVar(i, j, k_mirror, HydroSystem<TheProblem>::density_index);
+	const double x1Mom = consVar(i, j, k_mirror, HydroSystem<TheProblem>::x1Momentum_index);
+	const double x2Mom = consVar(i, j, k_mirror, HydroSystem<TheProblem>::x2Momentum_index);
+	double x3Mom = consVar(i, j, k_mirror, HydroSystem<TheProblem>::x3Momentum_index);
+	const double etot = consVar(i, j, k_mirror, HydroSystem<TheProblem>::energy_index);
+	const double eint = consVar(i, j, k_mirror, HydroSystem<TheProblem>::internalEnergy_index);
```
Comment (BenWibking) [inline src/problems/TallBoxSf/testTallBoxSf.cpp]: Using `k_mirror` is correct when the 'prevent inflow' condition is triggered, but wrong when it's an outflow. When it's an outflow, it should just be `foextrap` using `khi` (or `klo`).

**Review by BenWibking — CHANGES_REQUESTED** [review]
I think there's still an issue with the boundary condition.

File: src/problems/TallBoxSf/testTallBoxSf.cpp (hunk 16 above)
Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: Oh, definitely. I forgot to update this. I have a separate PR for this that I didn't merge yet.

Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: This is correct, right? https://github.com/quokka-astro/quokka/blob/e550372e43f4303542a0cd776e00e91f48f02e61/src/simulation.hpp#L2503

Comment (BenWibking) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: Yes, I think that's right.

Comment (chongchonghe) [reply src/problems/TallBoxSf/testTallBoxSf.cpp]: I have merged it in this branch. Ready for review again.

**Review by BenWibking — APPROVED** [review]
